src/groth16/circuit/builder/mod.rs

```diff
@@ -287,27 +645,333 @@ where
 
-        while literal != 0 {
-            let wire = match literal % 2 {
-                0 => self.zero_wire(),
-                1 => self.unity_wire(),
-                _ => unreachable!(),
+    // }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////// Keccak Functions ////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////
+
+    fn keccakf_1600(&mut self, a: &mut [Word64; 25]) {
+        for i in 0..24 {
+            let mut array: [Word64; 5] = [Word64::default(); 5];
+
+            // Theta
+            unroll! {
+                for x in 0..5 {
+                    unroll! {
+                        for y_count in 0..5 {
+                            let y = y_count * 5;
+                            array[x] = self.u64_bitwise_op(&array[x], &a[x + y], Circuit::new_xor);
+                        }
+                    }
+                }
+            }
+
+            unroll! {
+                for x in 0..5 {
+                    unroll! {
+                        for y_count in 0..5 {
+                            let y = y_count * 5;
+                            a[y + x] = self.u64_fan_in([a[y + x], array[(x + 4) % 5],
+                                rotate_word64_left(array[(x + 1) % 5], 1)].iter(), Circuit::new_xor);
+                        }
+                    }
+                }
+            }
+
+            // Rho and pi
+            let mut _last = a[1];
+            unroll! {
+                for x in 0..24 {
+                    array[0] = a[PI[x]];
+                    a[PI[x]] = rotate_word64_left(_last, RHO[x]);
+                    _last = array[0];
+                }
+            }
+
+            // Chi
+            unroll! {
+                for y_step in 0..5 {
+                    let y = y_step * 5;
+
+                    unroll! {
+                        for x in 0..5 {
+                            array[x] = a[y + x];
+                        }
+                    }
+
+                    unroll! {
+                        for x in 0..5 {
+                            let not = self.u64_unary_op(&array[(x + 1) % 5], Circuit::new_not);
+                            let and = self.u64_bitwise_op(&not, &(array[(x + 2) % 5]), Circuit::new_and);
+                            a[y + x] = self.u64_bitwise_op(&array[x], &and, Circuit::new_xor);
+
+                        }
+                    }
+                }
             };
 
-            bits.push(wire);
-            literal >>= 1;
+            // Iota
+            let rc_num = self.const_word64(RC[i]);
+            a[0] = self.u64_bitwise_op(&a[0], &rc_num, Circuit::new_xor);
+        }
+    }
+
+    fn squeeze(&mut self, keccak: &mut KeccakInternal, output: &mut [Word8]) {
+        fn setout(src: &[Word8], dst: &mut [Word8], len: usize) {
```

Comment:
Is there a good reason that this function is defined inside of this function instead of outside in the usual scope?

Reply:
Yes, `setout` should never be used outside of squeeze since it literally is designed to A) be the same a tiny_keccak B) work because of the way it is used in `squeeze`. It would be very painful to use and easy to mess up.